### PR TITLE
EID-1117 - Turn DuplicateAssertionValidator into interface

### PR DESCRIPTION
**Pull request by JHjava**
- So it can be used with a different implementation in proxy node

Co-authored-by: Simon Worthington <simon.worthington@digital.cabinet-office.gov.uk>

### CODE REVIEW

File: hub-saml/src/main/java/uk/gov/ida/saml/core/validators/assertion/DuplicateAssertionValidatorImpl.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    private final ConcurrentMap<String, DateTime> duplicateIds;
+
+    @Inject
+    public DuplicateAssertionValidatorImpl(ConcurrentMap<String, DateTime> duplicateIds) {
+        this.duplicateIds = duplicateIds;
+    }
+
+    @Override
+    public void validateAuthnStatementAssertion(Assertion assertion) {
+        if (!valid(assertion)) throw new SamlValidationException(authnStatementAlreadyReceived(assertion.getID()));
```
Comment (tomrosier) [inline hub-saml/src/main/java/uk/gov/ida/saml/core/validators/assertion/DuplicateAssertionValidatorImpl.java]: Drop the throws onto a new line 😄

File: hub-saml/src/main/java/uk/gov/ida/saml/core/validators/assertion/DuplicateAssertionValidatorImpl.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    @Override
+    public void validateMatchingDataSetAssertion(Assertion assertion, String responseIssuerId) {
+        if (!valid(assertion)) throw new SamlValidationException(duplicateMatchingDataset(assertion.getID(), responseIssuerId));
```
Comment (tomrosier) [inline hub-saml/src/main/java/uk/gov/ida/saml/core/validators/assertion/DuplicateAssertionValidatorImpl.java]: Drop throws onto a new line

File: hub-saml/src/main/java/uk/gov/ida/saml/core/validators/assertion/DuplicateAssertionValidatorImpl.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    private boolean valid(Assertion assertion) {
+        if (isDuplicateNonExpired(assertion)) return false;
```
Comment (tomrosier) [inline hub-saml/src/main/java/uk/gov/ida/saml/core/validators/assertion/DuplicateAssertionValidatorImpl.java]: Drop return onto new line

File: hub-saml/src/main/java/uk/gov/ida/saml/core/validators/assertion/DuplicateAssertionValidatorImpl.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+        DateTime expire = assertion.getSubject().getSubjectConfirmations().get(0).getSubjectConfirmationData().getNotOnOrAfter();
```
Comment (tomrosier) [inline hub-saml/src/main/java/uk/gov/ida/saml/core/validators/assertion/DuplicateAssertionValidatorImpl.java]: This is a very long line, what happens if any of the elements are null?

Comment (JHjava) [reply hub-saml/src/main/java/uk/gov/ida/saml/core/validators/assertion/DuplicateAssertionValidatorImpl.java]: I believe they are checked before this is called.

**Comment by Wynndow** [comment]
I'm struggling to understand the usefulness of this. I thought it might be because it allowed you to give a different implementation of `duplicateAssertionValidator` to `AuthnStatementAssertionValidator` - which would have made sense to me if `AuthnStatementAssertionValidator` was shared between the hub and the proxy-node - but it isn't. Am I missing something?

**Pull request merged**